storage/flush.go

```diff
@@ -21,157 +21,132 @@
+}
 
-	fm.Lock()
-	if fm.fs == flushInProgress {
-		fm.Unlock()
-		return
+func (mgr *flushManager) Flush(t time.Time) error {
+	timesToFlush := mgr.flushTimes(t)
+	if len(timesToFlush) == 0 {
+		return nil
 	}
-	fm.fs = flushInProgress
-	fm.Unlock()
-
-	flushFn := func() {
-		ctx := fm.opts.GetContextPool().Get()
-		defer ctx.Close()
-
-		for _, flushTime := range timesToFlush {
-			success := fm.flushWithTime(ctx, flushTime)
-			fm.Lock()
-			flushState := fm.flushAttempted[flushTime]
-			if success {
-				flushState.Status = flushSuccess
-			} else {
-				flushState.Status = flushFailed
-				flushState.NumFailures++
-			}
-			fm.flushAttempted[flushTime] = flushState
-			fm.Unlock()
+	ctx := mgr.opts.GetContextPool().Get()
+	defer ctx.Close()
+
+	multiErr := xerrors.NewMultiError()
+	for _, flushTime := range timesToFlush {
+		mgr.Lock()
+		if !mgr.needsFlush(flushTime) {
+			continue
 		}
-
-		fm.Lock()
-		fm.fs = flushNotStarted
-		fm.Unlock()
+		flushState := mgr.flushStates[flushTime]
+		flushState.Status = fileOpInProgress
+		mgr.flushStates[flushTime] = flushState
+		mgr.Unlock()
+
+		flushErr := mgr.flushWithTime(ctx, flushTime)
+
+		mgr.Lock()
+		flushState = mgr.flushStates[flushTime]
+		if flushErr == nil {
+			flushState.Status = fileOpSuccess
+		} else {
+			flushState.Status = fileOpFailed
+			flushState.NumFailures++
+			multiErr = multiErr.Add(flushErr)
+		}
+		mgr.flushStates[flushTime] = flushState
+		mgr.Unlock()
 	}
+	return multiErr.FinalError()
+}
+
+// flushTimes returns a list of times we need to flush data blocks for.
+func (mgr *flushManager) flushTimes(t time.Time) []time.Time {
+	earliest, latest := mgr.FlushTimeStart(t), mgr.FlushTimeEnd(t)
 
-	if !async {
-		flushFn()
-	} else {
-		go flushFn()
+	// NB(xichen): could preallocate slice here.
+	var flushTimes []time.Time
+	mgr.RLock()
+	for flushTime := latest; !flushTime.Before(earliest); flushTime = flushTime.Add(-mgr.blockSize) {
+		if mgr.needsFlush(flushTime) {
+			flushTimes = append(flushTimes, flushTime)
+		}
 	}
+	mgr.RUnlock()
+
+	return flushTimes
 }
 
-func (fm *flushManager) getTimesToFlush(t time.Time) []time.Time {
-	rops := fm.opts.GetRetentionOptions()
-	blockSize := rops.GetBlockSize()
-	maxFlushRetries := fm.opts.GetMaxFlushRetries()
-	firstBlockStart := fm.getFirstBlockStart(t)
-	earliestTime := t.Add(-1 * rops.GetRetentionPeriod())
-
-	fm.Lock()
-	defer fm.Unlock()
-	fm.flushTimes = fm.flushTimes[:0]
-	for flushTime := firstBlockStart; !flushTime.Before(earliestTime); flushTime = flushTime.Add(-blockSize) {
-		if flushState, exists := fm.flushAttempted[flushTime]; !exists || (flushState.Status == flushFailed && flushState.NumFailures < maxFlushRetries) {
-			flushState.Status = flushInProgress
-			fm.flushTimes = append(fm.flushTimes, flushTime)
-			fm.flushAttempted[flushTime] = flushState
-		}
+// needsFlush returns true if we need to flush data for a given time.
+func (mgr *flushManager) needsFlush(t time.Time) bool {
```

Comment:
Maybe leave in the comment that this needs to have the mgr read lock before calling needsFlush?